### ISSUE TITLE
Emit an error when external modules and Gleam modules would collide

### ISSUE THREAD

**Issue by PaperLightbulb** [gleam-lang/gleam#2738]
To replicate:
- create a gleam program

- create a mjs file with the same name (excluding the .mjs/.gleam extension) as one of your gleam files
OR
- create an erlang file with the same name as one of your gleam files

- attempt to import an external gleam function from one of the external files with the same name

Problem:
Gleam will not recognize any exports from the .mjs or .erl and throw the following error:
SyntaxError: The requested module '(mjs/erlang file path)' does not provide an export named '(imported external function name)'
It would be much more clear if the error was specific to having a mjs or erlang file with the same name as a gleam file importing from it, instead of stating that the module does not have the specified export

**Comment by lpil** [comment]
Thank you

**Comment by louiidev** [comment]
Hey @lpil do you mind if I pick up this issue?

**Comment by lpil** [comment]
Go for it! Thank you

**Comment by PastMoments** [comment]
I've just ran into this issue with using `@external(erlang, "zlib", "open")`, and naturally naming the file `zlib.gleam`. The error is even more unclear in this case, where it just says "function open/0 undefined". Definitely interested in seeing this resolved.

**Comment by giacomocavalieri** [comment]
@louiidev are you still planning to work on this? Otherwise I could try and pick it up

**Comment by louiidev** [comment]
> @louiidev are you still planning to work on this? Otherwise I could try and pick it up

Go ahead

**Comment by PgBiel** [comment]
@lpil I believe this can be closed since https://github.com/gleam-lang/gleam/pull/3601 was merged, which detects clashing `.gleam` and `.mjs` in the same directory and `.gleam` and `.erl` in any directories

Although I believe the case mentioned in https://github.com/gleam-lang/gleam/issues/2738#issuecomment-2053796359 (also considering dependencies) isn't covered, but maybe we can open a separate issue for that?

**Comment by PgBiel** [comment]
Actually it's worth mentioning that I sadly couldn't restrict clashing `.gleam` and `.erl` due to the way a Gleam project is packaged to Hex (which includes `.erl` files with the same name). I wonder if something could be done about that, but I had to leave that off unfortunately.

**Comment by lpil** [comment]
We would not run this for anything but top level packages.

**Comment by PgBiel** [comment]
Good point. Perhaps we could just restrict this check to when we're compiling the root project. Although that opens the possibility of libraries breaking unexpectedly, but if they run unit tests before publishing _like they should_, then that shouldn't be a problem.

However, I wonder if this could also affect something like `gleam run -m submodule`? If the root project is shifted to `submodule` and it comes from Hex, it's possible the check could accidentally be triggered. But I assume we still retain the original project information somewhere?

**Comment by lpil** [comment]
> Although that opens the possibility of libraries breaking unexpectedly

How would that happen? A package that fails to compile cannot be published to Hex.

> However, I wonder if this could also affect something like gleam run -m submodule? If the root project is shifted to submodule and it comes from Hex, it's possible the check could accidentally be triggered. But I assume we still retain the original project information somewhere?

Dependencies do not become the project root for `gleam run -m ...`, so it would not happen.
